### PR TITLE
KN-42465 Download from RedShift using Unload feature

### PR DESCRIPTION
**Pull request by BillSytin**
done

### CODE REVIEW

File: Knoema.Client/Client.cs [hunk 1]
```diff
@@ -18,343 +19,540 @@
...
+			}
+
+			return streamTasks;
+		}
+
+		public async Task<string[]> UnloadToLocalFolder(PivotRequest request, string destinationFolder)
+		{
+			var unloadData = await UnloadData(request);
+
+			var files = unloadData.Files.ToArray();
```
Comment (PavelPimenov) [inline Knoema.Client/Client.cs]: Why do you need to convert ToArray here it is List, why not just use it?

File: Knoema.Client/Client.cs [hunk 2]
```diff
@@ -18,343 +19,540 @@
...
+			if (taskResult.Status == Meta.TaskStatus.Failed)
+				throw new Exception(taskResult.Message);
+
+			if (taskResult.Status == Meta.TaskStatus.Completed)
+				return taskResult;
+
+			throw new ArgumentOutOfRangeException("Unexpected task status");
+		}
+
+		public Task<UnloadResponse> Unload(PivotRequest request)
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Looks like it should be private.

File: Knoema.Client/Client.cs [hunk 3]
```diff
@@ -18,343 +19,540 @@
...
+		}
+
+		public async Task<DatasetUnloadTaskResultData> UnloadData(PivotRequest request, int spinDelayInSeconds = 10, int maxWaitCount = 360)
+		{
+			var unloadTask = await Unload(request);
+			var taskResult = await WaitTaskResult<DatasetUnloadTaskResult>(unloadTask.TaskKey, spinDelayInSeconds, maxWaitCount);
+			return taskResult.Data;
+		}
+
+		public Task<Task>[] UnloadGetFiles(string[] files, Stream[] resultStreams)
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Looks like an internal private methoud as well.
Strange name by the way.

File: Knoema.Client/Client.cs [hunk 4]
```diff
@@ -18,343 +19,540 @@
...
+							response = t.GetAwaiter().GetResult();
+							return response;
+						}
+						catch (Exception)
+						{
+							cts.Cancel();
+							if (response != null)
+								response.Dispose();
+
+							throw;
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Does the exception go to the top level (out of "ContinueWith" body)? If not, then you have to check the token for cancellation before next loop.

File: Knoema.Client/Client.cs [hunk 5]
```diff
@@ -18,343 +19,540 @@
...
+			catch (Exception)
+			{
+				cts.Cancel();
+				throw;
+			}
+
+			return streamTasks;
+		}
+
+		public async Task<string[]> UnloadToLocalFolder(PivotRequest request, string destinationFolder)
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Give it the same "UnloadData" name.

File: Knoema.Client/Client.cs [hunk 6]
```diff
@@ -18,343 +19,540 @@
...
+				{
+					for (var i = 0; i < fileNames.Length; i++)
+					{
+						if (!string.IsNullOrEmpty(fileNames[i]))
+						{
+							try
+							{
+								File.Delete(destinationFolder + '\\' + fileNames[i]);
+							}
+							catch (Exception)
```
Comment (unicrus) [inline Knoema.Client/Client.cs]: Just "catch { }"

File: Knoema.Client/Data/UnloadResult.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (unicrus) [inline Knoema.Client/Data/UnloadResult.cs]: Remove redundant usings.

File: Knoema.Client/Meta/DatasetUnloadTaskResultData.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (unicrus) [inline Knoema.Client/Meta/DatasetUnloadTaskResultData.cs]: Remove redundant usings.

File: Knoema.Client/Client.cs (hunk 2 above)
Comment (BillSytin) [reply Knoema.Client/Client.cs]: ok

File: Knoema.Client/Client.cs (hunk 3 above)
Comment (BillSytin) [reply Knoema.Client/Client.cs]: ok, renamed

File: Knoema.Client/Client.cs (hunk 4 above)
Comment (BillSytin) [reply Knoema.Client/Client.cs]: it should, why not?

File: Knoema.Client/Client.cs (hunk 5 above)
Comment (BillSytin) [reply Knoema.Client/Client.cs]: renamed 'UnloadData' to 'Unload'

File: Knoema.Client/Client.cs (hunk 1 above)
Comment (BillSytin) [inline Knoema.Client/Client.cs]: It's just to ensure immutable length and to be the same as the arrays below.

File: Knoema.Client/Client.cs (hunk 6 above)
Comment (BillSytin) [reply Knoema.Client/Client.cs]: ok

File: Knoema.Client/Data/UnloadResult.cs (hunk 7 above)
Comment (BillSytin) [reply Knoema.Client/Data/UnloadResult.cs]: ok

File: Knoema.Client/Meta/DatasetUnloadTaskResultData.cs (hunk 8 above)
Comment (BillSytin) [reply Knoema.Client/Meta/DatasetUnloadTaskResultData.cs]: ok

**Pull request merged**